File: src/windows/AmbientContextMcp.Core/Hub/PolicyVersionService.cs
using System.Security.Cryptography;
using System.Text;
using AmbientContextMcp.Core.Models;

namespace AmbientContextMcp.Core.Hub;

public static class PolicyVersionService
{
    public static string ComputePolicyVersion(
        IReadOnlyList<PrivacyClassification> classifications,
        IReadOnlyDictionary<string, bool> overrides)
    {
        var sb = new StringBuilder();

        foreach (var item in classifications.OrderBy(c => c.Path, StringComparer.OrdinalIgnoreCase))
        {
            sb.Append("c|").Append(item.Path).Append('|')
              .Append(SensitivityScopeFilter.NormalizeSensitivity(item.Sensitivity)).Append('|')
              .Append(item.DefaultTransmit ? '1' : '0').Append('\n');
        }

        foreach (var pair in overrides.OrderBy(p => p.Key, StringComparer.OrdinalIgnoreCase))
        {
            sb.Append("o|").Append(pair.Key).Append('|')
              .Append(pair.Value ? '1' : '0').Append('\n');
        }

        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(sb.ToString()));
        return Convert.ToBase64String(hash, 0, 9)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }
}

File: src/macos/Sources/AmbientContextCore/Hub/PolicyVersionService.swift
import CryptoKit
import Foundation

public enum PolicyVersionService {
    /// 分類 (path 昇順) と override (key 昇順) を連結した文字列の SHA-256 先頭 9 バイトを
    /// base64url (パディング無し) にしたもの。C# 実装とバイト単位で一致する。
    public static func computePolicyVersion(
        classifications: [PrivacyClassification],
        overrides: CaseInsensitiveDictionary<Bool>
    ) -> String {
        var text = ""

        for item in classifications.sorted(by: { $0.path.lowercased() < $1.path.lowercased() }) {
            text += "c|" + item.path + "|"
                + SensitivityScopeFilter.normalizeSensitivity(item.sensitivity) + "|"
                + (item.defaultTransmit ? "1" : "0") + "\n"
        }

        for pair in overrides.sortedPairs {
            text += "o|" + pair.key + "|" + (pair.value ? "1" : "0") + "\n"
        }

        let digest = SHA256.hash(data: Data(text.utf8))
        let prefix = Data(Array(digest).prefix(9))
        return Base64Url.encode(prefix)
    }
}
